### PR TITLE
[#47629] Prepare the database to store Nextcloud application password and managed folder flag.

### PR DESCRIPTION
**Pull request by ba1ash**
https://community.openproject.org/work_packages/47629

- Turn `storages.provider_type` into a column for STI.
- Introduce `storages.provider_fileds` column which is responsible for storing different attributes depending on a specific storage provider. For nextcloud there are `username`, `password`, `managed_folders`.
- Introduce new storage model: `Storages::NextcloudStorage`.
- Migrate old code to respect new model.
- Remove a dead link from the code base: https://www.openproject.org/docs/development/concepts/contracted-services/

### CODE REVIEW

File: db/migrate/20230420071113_migrate_storages_to_use_provider_type_as_sti_column.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class MigrateStoragesToUseProviderTypeAsStiColumn < ActiveRecord::Migration[7.0]
+  def up
+    Storages::Storage.update_all("provider_type = 'Storages::NextcloudStorage'")
```
Comment (ba1ash) [inline db/migrate/20230420071113_migrate_storages_to_use_provider_type_as_sti_column.rb]: Add a `where` condition for change only `nextcloud` rows.

File: modules/storages/app/models/storages/nextcloud_storage.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+class Storages::NextcloudStorage < Storages::Storage
+  store :provider_fields, accessors: %i[username password managed_folders]
+
+  def managed_folders=(value)
+    super(!!value)
+  end
+
+  def managed_folders
```
Comment (ba1ash) [inline modules/storages/app/models/storages/nextcloud_storage.rb]: Let's rename to `has_managed_project_folders`.

File: modules/storages/app/models/storages/storage.rb [hunk 3]
```diff
@@ -77,4 +79,12 @@ class Storages::Storage < ApplicationRecord
       )
     end
   }
+
+  def self.shorten_provider_type(provider_type)
+    provider_type.demodulize[0..-8].downcase
```
Comment (ba1ash) [inline modules/storages/app/models/storages/storage.rb]: Let's get rid of magic numbers and rewrite the code to depend on the `Storage` string explicitly.

File: modules/storages/app/models/storages/nextcloud_storage.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Storages::NextcloudStorage < Storages::Storage
+  store :provider_fields, accessors: %i[username password has_managed_project_folders]
+
+  def has_managed_project_folders=(value)
+    super(!!value)
+  end
+
+  def has_managed_project_folders
```
Comment (wielinde) [inline modules/storages/app/models/storages/nextcloud_storage.rb:11]: @ba1ash Even rubocop wants this method to be a predicate with a question mark at the end :-)

Comment (ba1ash) [reply modules/storages/app/models/storages/nextcloud_storage.rb:11]: True :)
I added an alias to the method.

File: modules/storages/app/models/storages/storage.rb [hunk 5]
```diff
@@ -77,4 +79,17 @@ class Storages::Storage < ApplicationRecord
       )
     end
   }
+
+  def self.shorten_provider_type(provider_type)
+    case /Storages::(?'storage_name'.*)Storage/.match(provider_type)
+    in storage_name:
```
Comment (wielinde) [inline modules/storages/app/models/storages/storage.rb]:
```suggestion
    in provider_name:
```

File: modules/storages/app/models/storages/storage.rb [hunk 6]
```diff
@@ -77,4 +79,18 @@ class Storages::Storage < ApplicationRecord
       )
     end
   }
+
+  def self.shorten_provider_type(provider_type)
+    case /Storages::(?'storage_name'.*)Storage/.match(provider_type)
```
Comment (ba1ash) [inline modules/storages/app/models/storages/storage.rb]:
```suggestion
    case /Storages::(?'provider_name'.*)Storage/.match(provider_type)
```

File: modules/storages/app/models/storages/storage.rb [hunk 7]
```diff
@@ -77,4 +79,18 @@ class Storages::Storage < ApplicationRecord
...
+    in storage_name:
+      storage_name.downcase
```
Comment (ba1ash) [inline modules/storages/app/models/storages/storage.rb]:
```suggestion
      provider_name.downcase
```

**Review by wielinde — APPROVED** [review]
LGTM. There is a smaller naming thing. After changing it please merge.

There is a big open topic how to handle different storage provider types in terms of services, contracts and forms. The PR here does not yet tackle these topics. We will need to work on those questions later.

**Pull request merged**